### PR TITLE
allow to select sha algorithm from config file

### PR DESCRIPTION
**Pull request by janciesla8818**
This PR extends functionality of multi sha support. This PR allows to specify which sha algorithms can be used for storing passwords in scram format. 

The change of the config file allows now to specify scram password format in two different ways:
* {password_format, scram} - how the password format was specified so far
* {password_format, {scram, [sha, sha256]}} - new format allowing to select sha algorithms within the list. This format will allow to store the scram of the passwords using only those algorithms that are specified in the list.

### CODE REVIEW

**Comment by janciesla8818** [comment]
Documentation changes that are needed due to this change are still pending. Will be updated shortly.

File: src/mongoose_scram.erl [hunk 1]
```diff
@@ -153,7 +155,19 @@ serialize(#{salt   := Salt, iteration_count := IterationCount,
...
-    <<?SCRAM_SHA256_PREFIX>>/binary, Sha256StoredKey/binary, $|, Sha256ServerKey/binary >>.
+    <<?SCRAM_SHA256_PREFIX>>/binary, Sha256StoredKey/binary, $|, Sha256ServerKey/binary >>;
+serialize(#{salt   := Salt, iteration_count := IterationCount,
+            sha    := #{server_key := ServerKey, stored_key := StoredKey}}) ->
+    IterationCountBin = integer_to_binary(IterationCount),
+    << <<?MULTI_SCRAM_SERIAL_PREFIX>>/binary,
+    Salt/binary, $,, IterationCountBin/binary, $,,
+    <<?SCRAM_SHA_PREFIX>>/binary, StoredKey/binary, $|, ServerKey/binary >>;
+serialize(#{salt   := Salt, iteration_count := IterationCount,
+            sha256 := #{server_key := ServerKey, stored_key := StoredKey}}) ->
```
[the remark this answers is not in the document]
Comment (janciesla8818) [reply src/mongoose_scram.erl:166]: I struggled a bit to write something readable and not repetitive using this binary syntax ... so I went with readable.

File: src/mongoose_scram.erl [hunk 2]
```diff
@@ -101,23 +101,25 @@ mask(Key, Data) ->
     <<C:KeySize>>.
 
 enabled(Host) ->
-    ejabberd_auth:get_opt(Host, password_format) == scram.
+    PasswordFormat = ejabberd_auth:get_opt(Host, password_format),
+    ScramSha = proplists:get_value(scram, [PasswordFormat]),
+    PasswordFormat == scram orelse PasswordFormat == {scram, ScramSha}.
```
Comment (michalwski) [inline src/mongoose_scram.erl:106]: It took me a while to understand what happens here. To my understanding this is equivalent of the following:
```suggestion
    case ejabberd_auth:get_opt(Host, password_format) of
        scram -> true;
        {scram, _ScramSha} -> true;
        _ -> false
    end.
```

File: src/mongoose_scram.erl [hunk 3]
```diff
@@ -230,3 +244,13 @@ do_check_digest([{Sha,_Prefix} | RemainingSha], ScramMap, Digest, DigestGen, Pas
...
+
+configured_sha_types(Host) ->
+    PasswordFormat = ejabberd_auth:get_opt(Host, password_format),
+    ScramSha = proplists:get_value(scram, [PasswordFormat]),
+    case PasswordFormat == scram orelse ScramSha == [] of
+        true -> supported_sha_types();
+        false ->
+            lists:filter(fun({Sha, _Prefix}) ->
+                            lists:member(Sha, ScramSha) end, supported_sha_types())
+    end.
```
Comment (michalwski) [inline src/mongoose_scram.erl:256]: Does the following do the same?
```suggestion
    case PasswordFormat = ejabberd_auth:get_opt(Host, password_format) of
        scram -> supported_sha_types();
        {scram, ScramSha} ->
            lists:filter(fun({Sha, _Prefix}) ->
                            lists:member(Sha, ScramSha) end, supported_sha_types())
    end.
```

File: src/mongoose_scram.erl (hunk 1 above)
[the remark this answers is not in the document]
Comment (michalwski) [reply src/mongoose_scram.erl:166]: I see 2 ways how to make this more generic:
1. Put all sha specific data in a map (or list) like below
```erlang
#{salt := <<"this is salt">>,
   iteration_count := 4503,
   scram := #{sha := #{server_key := <<"server key foor sha1">>,
                       stored_key := <<"stored key for sha2">>},
              sha256 := #{server_key := <<"server key for sha256">>,
                          stored_key := <<"stored key for sha256">>}
   }
}
```
And than iterate over what's inside the `scram` key and serialise accordingly.

2. Use 
```erlang
maps:map(fun serialise_a_sha/2, maps:with(supported_sha_types(), TheMap))
```

File: src/mongoose_scram.erl [hunk 4]
```diff
@@ -101,23 +101,28 @@ mask(Key, Data) ->
     <<C:KeySize>>.
 
 enabled(Host) ->
-    ejabberd_auth:get_opt(Host, password_format) == scram.
+    case ejabberd_auth:get_opt(Host, password_format) of
+        scram -> true;
+        {scram} -> true;
```
Comment (michalwski) [inline src/mongoose_scram.erl:106]: Why do we need this clause?

File: src/mongoose_scram.erl [hunk 5]
```diff
@@ -230,3 +242,16 @@ do_check_digest([{Sha,_Prefix} | RemainingSha], ScramMap, Digest, DigestGen, Pas
 supported_sha_types() ->
     [{sha,      <<?SCRAM_SHA_PREFIX>>},
      {sha256,   <<?SCRAM_SHA256_PREFIX>>}].
+
+configured_sha_types(Host) ->
+    PasswordFormat = ejabberd_auth:get_opt(Host, password_format),
+    ScramSha = proplists:get_value(scram, [PasswordFormat]),
+    case ejabberd_auth:get_opt(Host, password_format) of
+        Scram when Scram == scram orelse Scram == {scram} ->
```
Comment (michalwski) [inline src/mongoose_scram.erl:250]: When is it possible to have `{scram}` as a value of the config option?

Comment (janciesla8818) [reply src/mongoose_scram.erl:250]: In general it should not be possible but I figured that it could be common typo when someone has {scram, [sha256]} and decides to go back to just scram leaving {scram}. It doesn't cost us much but the config would be more forgiving with the syntax typo.

File: doc/Advanced-configuration.md [hunk 6]
```diff
@@ -140,12 +140,12 @@ Retaining the default layout is recommended so that the experienced MongooseIM u
...
+    * **Description:** Provides different parameters that will be applied to a chosen authentication method.
                        `auth_password_format` and `auth_scram_iterations` are common to `http`, `rdbms`, `internal` and `riak`.
 
         * **auth_password_format**
-             * **Description:** Decide whether user passwords will be kept plain or hashed in the database. Currently the popular XMPP clients support the SCRAM method, so it is strongly recommended to use the hashed version. The older ones can still use `PLAIN` mechiansm. `DIGEST-MD5` is not available with `scram`.
-             * **Values:** `plain`, `scram`
+             * **Description:** Decide whether user passwords will be kept plain or hashed in the database. Currently, the popular XMPP clients support the SCRAM method. It is strongly recommended to use the hashed version. SCRAM hashing with use of the SHA-1 and SHA-256 is supported. Hashing algorithms can be provided as an argument. This will result in storing and supporting only hashes specified in the configuration. The older XMPP clients can still use `PLAIN` mechanism. `DIGEST-MD5` is not available with `scram`.
```
Comment (goddammit) [inline doc/Advanced-configuration.md:147]:
```suggestion
             * **Description:** Decide whether user passwords will be kept plain or hashed in the database. Currently, popular XMPP clients support the SCRAM method and it is strongly recommended to use the hashed version. MongooseIM supports SCRAM hashing, either SHA-1 or SHA-256 can be provided as an argument and this will result in storing and supporting only hashes specified in the configuration. The older XMPP clients can still use the `PLAIN` mechanism. `DIGEST-MD5` is not available with `scram`.
```

???? I might be wrong on this, it's just that the flurry of short sentences sounds a bit robotic

File: doc/developers-guide/SCRAM-serialization.md [hunk 7]
```diff
@@ -5,6 +5,37 @@ Developers can use this information to create advanced endpoints for `ejabberd_a
...
 ## Format description
 
+`==MULTI_SCRAM==,<salt>,<iteration count>,===SHA1===<stored key>|<server key>,==SHA256==<stored key>|<server key>`
+
+* `<salt>` - Base64-encoded Salt
+* `<iteration count>` - Iteration Count formatted as a human-readable integer
+* `<stored key>` - Base64-encoded Stored Key
+* `<server key>` - Base64-encoded Server Key
+
+The SCRAM format can vary depending on the SHA algorithms that are used for SCRAM. Salt and iteration count is common for different SHA types. Stored Key and Server Key are specific to a given SHA and are following SHA prefix that is indicating which SHA they belong to.
```
Comment (goddammit) [inline doc/developers-guide/SCRAM-serialization.md:15]:
```suggestion
The SCRAM format can vary depending on the SHA algorithms that are used for SCRAM. Salt and iteration count is common for different SHA types. Stored Key and Server Key are specific to a given SHA and are following a SHA prefix that is indicating which SHA they belong to.
```

File: doc/developers-guide/SCRAM-serialization.md [hunk 8]
```diff
@@ -5,6 +5,37 @@ Developers can use this information to create advanced endpoints for `ejabberd_a
...
+        stored_key => <<"Hxfegcj54A6p/ZzWkw7BjJsLZpM=">>},
+   sha256 =>
+      #{server_key => <<"0Tm46vWQyO/XWpAtcDCUH++7ImWVDe0VDx274zMzMXQ=">>,
+        stored_key => <<"S4XiSkYI81m688nMuCgFGUBIGbW6oG1a9rlTKktzS/A=">>}}
+```
+* *Serialized password:* `==MULTI_SCRAM==,glF/hUXiCWD6TaJLpgVaaw==,4096,===SHA1===Hxfegcj54A6p/ZzWkw7BjJsLZpM=|ztvci4eBktO3NLviLhwN3ktF15Q=,==SHA256==S4XiSkYI81m688nMuCgFGUBIGbW6oG1a9rlTKktzS/A=|0Tm46vWQyO/XWpAtcDCUH++7ImWVDe0VDx274zMzMXQ=`
+
+## Legacy format description
+
+MongooseIM installations prior to 3.6.3 were supporting only SHA-1 as hashing algorithm for SCRAM. The SCRAM format that was used can be seen below.
```
Comment (goddammit) [inline doc/developers-guide/SCRAM-serialization.md:37]:
```suggestion
MongooseIM installations prior to 3.6.3 were supporting only SHA-1 as a hashing algorithm for SCRAM. The SCRAM format that was used can be seen below.
```

File: big_tests/tests/login_SUITE.erl [hunk 9]
```diff
@@ -39,20 +39,25 @@ all() ->
...
+         {configure_sha256, [pallarel], configure_sha256_test()},
          {login_scram_store_plain, [parallel], scram_tests()},
          {messages, [sequence], [messages_story, message_zlib_limit]}],
     ct_helper:repeat_all_until_all_ok(G).
 
 scram_tests() ->
     [log_one, log_one_scram, log_one_scram_256].
 
+configure_sha256_test() ->
+    [configure_sha256_log_with_sha256].
```
Comment (michalwski) [inline big_tests/tests/login_SUITE.erl:59]: I think it'd be beneficial to have a test case `configure_sha256_log_with_sha1_fails`, what do you think? This could be added in a separate PR, though.

**Comment by michalwski** [comment]
I think the following corner case is still possible:
* configured `{sasl_mechanisms, [cyrsasl_scram, cyrsasl_scram256]}`
* configured `{auth_password_format, {scram, [sha256]}}`

MogooseIM offers `SCRAM-SHA-1` and `SCRAM-SHA-256` as possible auth mechanism. The user chooses `SCRAM-SHA-1` but the authentication fails since `sha1` is no on the `auth_password_format`.
